new-tutorial/3_message_passing.py

```diff
@@ -0,0 +1,353 @@
+
+"""
+
+import dgl
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+######################################################################
+# This tutorial assumes that you already know the pipeline of full-graph
+# node classification. If not, please refer to the
+# :doc:`introduction <1_introduction>`.
+# 
+# The following code for data loading and training loop is directly copied
+# from the introduction tutorial.
+# 
+
+import dgl.data
+
+dataset = dgl.data.CoraGraphDataset()
+g = dataset[0]
+
+def train(g, net):
+    optimizer = torch.optim.Adam(net.parameters(), lr=0.01)
+    all_logits = []
+    best_val_acc = 0
+    best_test_acc = 0
+
+    features = g.ndata['feat']
+    labels = g.ndata['label']
+    train_mask = g.ndata['train_mask']
+    val_mask = g.ndata['val_mask']
+    test_mask = g.ndata['test_mask']
+    for e in range(200):
+        # Forward
+        logits = net(g, features)
+
+        # Compute prediction
+        pred = logits.argmax(1)
+
+        # Compute loss
+        # Note that we should only compute the losses of the nodes in the training set,
+        # i.e. with train_mask 1.
+        loss = F.cross_entropy(logits[train_mask], labels[train_mask])
+
+        # Compute accuracy on training/validation/test
+        train_acc = (pred[train_mask] == labels[train_mask]).float().mean()
+        val_acc = (pred[val_mask] == labels[val_mask]).float().mean()
+        test_acc = (pred[test_mask] == labels[test_mask]).float().mean()
+
+        # Save the best validation accuracy and the corresponding test accuracy.
+        if best_val_acc < val_acc:
+            best_val_acc = val_acc
+            best_test_acc = test_acc
+
+        # Backward
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+        all_logits.append(logits.detach())
+
+        if e % 5 == 0:
+            print('In epoch {}, loss: {:.3f}, val acc: {:.3f} (best {:.3f}), test acc: {:.3f} (best {:.3f})'.format(
+                e, loss, val_acc, best_val_acc, test_acc, best_test_acc))
+
+
+######################################################################
+# Message passing and GNNs
+# ------------------------
+# 
+# DGL follows the *message passing paradigm* inspired by the Message
+# Passing Neural Network proposed by `Gilmer et
+# al. <https://arxiv.org/abs/1704.01212>`__ Essentially, they found many
+# GNN models can fit into the following framework:
+# 
+# .. math::
+# 
+# 
+#    m_{u\to v}^{(l)} = M^{(l)}\left(h_v^{(l-1)}, h_u^{(l-1)}, e_{u\to v}^{(l-1)}\right)
+# 
+# .. math::
+# 
+# 
+#    m_{v}^{(l)} = \sum_{u\in\mathcal{N}(v)}m_{u\to v}^{(l)}
+# 
+# .. math::
+# 
+# 
+#    h_v^{(l)} = U^{(l)}\left(h_v^{(l-1)}, m_v^{(l)}\right)
+# 
+# where DGL calls :math:`M^{(l)}` the *message function*, :math:`\sum` the
+# *reduce function* and :math:`U^{(l)}` the *update function*. Note that
+# :math:`\sum` here can represent any function and is not necessarily a
+# summation.
```

Comment:
Shall we use the same notation as in our paper?

Reply:
The notations are from the MPNN paper.  Since we are referencing their paper I prefer this notation.